### PR TITLE
Add dtype attribute to CSEVariable

### PR DESCRIPTION
**Pull request by arui-meta**
Summary:
- This diff introduces `dtype` attribute to `TritonCSEVariable` and a dtype propagation helper function to infer dtype from input to output for each op.

- There will be a follow-up diff that uses this `dtype` information in `TritonCSEVariable` to perform dtype-aware codegen.

Test Plan: CI

Differential Revision: D61815079

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

**Review by eellison — COMMENTED** [review]
Left a few comments - I think there is one general discussion that I tagged @blaine-rister on we should discuss.

To correctly propagate we'll need to track `is_scalar` as well - i don't think that would be too tricky, it's basically just the `constant` that should be a scalar i think. 

So long as we're not actually using the dtypes yet in inductor codegen we could potentially land this in steps but I do there's a bit more work before we can rely on this.

File: torch/_inductor/codegen/common.py
Comment (arui-meta) [reply torch/_inductor/codegen/common.py:1668]: Thanks! Will do!

File: torch/_inductor/codegen/triton.py
Comment (arui-meta) [reply torch/_inductor/codegen/triton.py:845]: Thanks for the feedback! 

It seems reasonable to always preserve the same dtype as torch by inserting upcast on input and downcast on output. We could also easily remove these casts when a given triton operator starts to support fp16/bf16 in the future.

**Comment by arui-meta** [comment]
In the latest version, I have made the following changes since last review

- Replace try-catch block with a default dtype propagation handler
- Updated dtype promotion logic to include scalars and use `torch._prims_common.elementwise_dtypes`
- Updated `torch._prims_common.elementwise_dtypes` to work with CSEVariable
- Updated dtype_aware_codegen test to test with both float16 and bfloat16 inputs
- Cast back to input dtype when triton function call automatically upcast to float32. 

I still kept the `promote_types` calls for unary operators because sometimes we receive a str on the input argument in template codegen. Will need a separate fix before we can rely on `arg.dtype`. 
@eellison @blaine-rister

File: test/inductor/test_torchinductor.py
Comment (blaine-rister) [inline test/inductor/test_torchinductor.py:12165]: Nice test. Does this PR already affect the codegen? From the description, it sounded like that would be a follow up.

File: torch/_inductor/codegen/common.py
Comment (blaine-rister) [reply torch/_inductor/codegen/common.py:1822]: Might be good to leave a TODO in the code so we don't forget :)

**Review by blaine-rister — COMMENTED** [review]
LGTM! I'll have to defer to @eellison  for the final approval. The basic logic makes sense to me. I didn't verify all the different dtype propagation rules for reductions, but it sounds like those will be tested in follow-up PRs. I left comments in a few places where we might be able to streamline the codegen, once we scale this up to cover more ops.

File: test/inductor/test_torchinductor.py
Comment (arui-meta) [reply test/inductor/test_torchinductor.py:12165]: We still keep the inductor flag `triton.codegen_upcast_to_fp32 = True` by default. Codegen will still see all CSEVariables in FP32 by default. I only updated the `sqrt` codegen so that I can put in an unit test to validate dtype-aware codegen works by switching off `triton.codegen_upcast_to_fp32` and feed in lower precision inputs.

File: torch/_prims_common/__init__.py
Comment (arui-meta) [reply torch/_prims_common/__init__.py:1516]: This util function has been updated to deal with CSEVariables, which do not have an attribute "ndim"

File: torch/_inductor/codegen/common.py
Comment (arui-meta) [reply torch/_inductor/codegen/common.py:1771]: Regarding scalar dtypes, the pytorch type promotion util function used here is`torch._prims_common.elementwise_dtypes` , which already handles scalars in input args. Depending on the scalar dtype relative to tensor dtype, this util function will decide whether or not scalar inputs will participate in dtype propagation. 

Regarding the triton util function, there is still an upcast to float32 in the promotion logic if the operator is div or mod. https://github.com/triton-lang/triton/blob/fa229d1c4bee16c094be9427334575ec1e79f66c/python/triton/language/semantic.py#L86

File: torch/_inductor/codegen/triton.py
Comment (arui-meta) [reply torch/_inductor/codegen/triton.py:853]: Great idea!

File: torch/_prims_common/__init__.py
Comment (eellison) [inline torch/_prims_common/__init__.py:1478]: We should not leak concepts from torchinductor into prims dtype promotion. You can avoid this by doing what inductor does [here]( https://github.com/pytorch/pytorch/blob/c272526ea54b8063c596addb48cd3a87ab91a5ef/torch/_inductor/lowering.py#L224).

File: torch/_inductor/codegen/triton.py
Comment (eellison) [inline torch/_inductor/codegen/triton.py:854]: There is a method for this elsewhere in this file, please reuse it. `triton_type`

File: torch/_inductor/codegen/triton.py
Comment (eellison) [inline torch/_inductor/codegen/triton.py:852]: Since the dtypes are not correct currently - we should use them in codegen yet

**Review by eellison — COMMENTED** [review]
Needs to fix the leaking of inductor into prims dtype promotion before landing, then we can land. 

It's make a bit nervous to have dtype promotion that is not totally correct in main. Can we guard any uses of the dtype in codegen with the config option? @blaine-rister commented on that as well.

Things that needs to be implemented still for correctness:

- using `is_scalar` 
- following `masked` 
- making sure that the triton types and pytorch types align - concretely, we would need to add casts whenever there is a divergence in codegen (such as with triton upcasting div/mod).

File: torch/_inductor/codegen/triton.py
Comment (arui-meta) [reply torch/_inductor/codegen/triton.py:854]: Right now it seems that `triton_type` will map float16 / bfloat16 to float32, which is not the original input dtype.

File: torch/_inductor/codegen/triton.py
Comment (arui-meta) [reply torch/_inductor/codegen/triton.py:852]: I see. The only reason I modified `sqrt` codegen is to have an unittest. I can remove it if it seems risky.

File: torch/_inductor/codegen/triton.py
Comment (eellison) [reply torch/_inductor/codegen/triton.py:854]: That's what `triton_compute_type` does. not `triton_type`.

File: torch/_inductor/codegen/triton.py
Comment (eellison) [reply torch/_inductor/codegen/triton.py:852]: Sorry, meant to type *should not. Let's still check `codegen_upcast_to_fp32` here just to be safe.

File: torch/_prims_common/__init__.py
Comment (eellison) [inline torch/_prims_common/__init__.py:1498]: Shouldn't need this change

File: torch/_inductor/codegen/common.py
Comment (eellison) [inline torch/_inductor/codegen/common.py:1776]: `torch.empty(0, dtype=inp.dtype)` actually has different type promotion rules than a tensor because it acts as a scalar. Can you do `torch.empty([1], dtype=inp.dtype)` instead ? Read the docs on type promotion here - https://pytorch.org/docs/stable/tensor_attributes.html#type-promotion-doc

` # torch.empty(0) has ndim == 1` delete this comment  ?

File: torch/_inductor/codegen/common.py
Comment (eellison) [inline torch/_inductor/codegen/common.py:1733]: Just commenting on this now but.. It's better practice to add a new arg at the end of arguments, typically...

**Review by eellison — APPROVED** [review]
please remove change to `_prims_common` and fix constructing 0-dim tensor for the type promotion

File: torch/_inductor/codegen/common.py
Comment (blaine-rister) [reply torch/_inductor/codegen/common.py:1788]: Got it! Thanks for clarifying. LGTM.

**Review by blaine-rister — APPROVED** [review]
LGTM! Thanks for the new feature.

**Pull request closed without merging**
